In [1]:
#Student ID 1002080520
import numpy as np
from numpy import loadtxt
import csv

#CODE TO TAKE THE DATA AND STORE IN THE VARIABLE

file = open('q2_cde_dataset.csv')
csvreader = csv.reader(file, delimiter = ",")
data = []
for row in csvreader:
    data.append(row)

program_data = []
for i in range(0 , len(data)):
    col= []
    for j in range(0,len(data[i])):
        if j==len(data[i])-1:
            col.append(data[i][j])
        elif j==3:
            j=j+1
        else:
            col.append(float(data[i][j]))
    program_data.append(col)
file.close()
print("\n\nProgram Data \n\n")
print(program_data)

#We have successfully imported the data without the fourth feature in given data file



Program Data 


[[0.088806749762916, 0.098008394324455, 0.57649058735898, ' Ceramic'], [0.078963683508211, 0.085300149997219, 0.27162441732486, ' Metal'], [0.1317769211421, 0.070414892852672, 0.50961174676178, ' Ceramic'], [0.12238449845956, 0.081946645569603, 0.64441296678112, ' Ceramic'], [0.11207208488382, 0.15, 0.3736066136156, ' Plastic'], [0.11696836534332, 0.1192966406841, 0.3315496717105, ' Plastic'], [0.12666802764538, 0.15, 0.41499442802485, ' Plastic'], [0.10444296796273, 0.10662125348196, 0.29055615215118, ' Plastic'], [0.13065405525064, 0.14162609670298, 0.33516279766318, ' Plastic'], [0.11931866537973, 0.10053257281629, 0.29777585695821, ' Plastic'], [0.082429457069664, 0.03, 0.13468307216616, ' Ceramic'], [0.091479158720172, 0.060070441860045, 0.19527332743138, ' Metal'], [0.071660045050914, 0.068186655985927, 0.20326252583045, ' Metal'], [0.076470494832748, 0.037424113599498, 0.14517371741585, ' Metal'], [0.071541180684419, 0.046722582104868, 0.14136615592985, ' Metal

In [5]:
#Training the program data by giving data itself and predicting
Train_program_data = program_data
Distance =[]
for i in range(0 , len(Train_program_data)):
    col = []
    for j in range(0,len(Train_program_data)):
        if i==j:
             j=j+1
        else:
                dist_label =[]
                dist_label.append(Train_program_data[j][3])
                X=Train_program_data[i][0]
                Y=Train_program_data[i][1]
                Z=Train_program_data[i][2]
                x=Train_program_data[j][0]
                y=Train_program_data[j][1]
                z=Train_program_data[j][2]
                dist = np.sqrt((X-x)**2 + (Y-y)**2 + (Z-z)**2)
                dist_label.append(dist)
                col.append(dist_label)
    Distance.append(col)
print("\n DISTANCE MATRIX \n")
print(Distance)
print("\n") 

In [8]:
#We have measured the distances and also labeled the points now we need to extract the distance arrange it into 
#ascending order and the again attached the labels so that we can get newly trained data set
def bubblesort(arr):   
    n = len(arr)
    # For loop to traverse through all
    # element in an array
    for i in range(n):
        for j in range(0, n - i - 1):
             
            # Range of the array is from 0 to n-i-1
            # Swap the elements if the element found
            #is greater than the adjacent element
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                
only_dist =[]
for i in range(0, len(Distance)):
    cols=[]
    for j in range(0 , len(Distance[i])):
        cols.append(Distance[i][j][1])
    bubblesort(cols)
    only_dist.append(cols)
print(only_dist)

[[0.029865896676656425, 0.041048196992840076, 0.049550178335158505, 0.05669553212060376, 0.062351416397756654, 0.07745253131411604, 0.07893535522645392, 0.08414639785193075, 0.0877469565348022, 0.09097338400492193, 0.0935992030723897, 0.09676575322852574, 0.10022965218728348, 0.1036376456365194, 0.10616868217993887, 0.10666004847623738, 0.11667741680561848, 0.13730604888485323, 0.14222063349252134, 0.1437238873943267, 0.14619991182839268, 0.15129098693453816, 0.15233336159291777, 0.15279408814449522, 0.15507917212563435, 0.1628694877379771, 0.16314926565961896, 0.16707981499114857, 0.1738321400181486, 0.17556669528340224, 0.1768441700287177, 0.18540768851659614, 0.18879695883879338, 0.19921357109201007, 0.20031906828952775, 0.2045812398144893, 0.20612507304483846, 0.20672744771313523, 0.20749977494817318, 0.20829671902036015, 0.21072804673539264, 0.21500301298295532, 0.217661643261342, 0.2353707936792588, 0.2370207628516403, 0.24007094578627422, 0.24701134627937957, 0.24747185331645935

In [9]:
#We have arranged the distance in ascending order now we will label it
data_dist_label=[]
for i in range(0 , len(only_dist)):
    cols=[]
    for j in range(0 , len(only_dist[i])):
        for h in range(0 , len(only_dist[i])):
            if only_dist[i][j]==Distance[i][h][1]:
                d = Distance[i][h][0]
                break
        cols.append(d)
    data_dist_label.append(cols)

print(data_dist_label)

#We have successfully generated the distance label in ascending order

[[' Metal', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Ceramic', ' Plastic', ' Metal', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Metal', ' Metal', ' Metal', ' Metal', ' Ceramic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Metal', ' Ceramic', ' Metal', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Ce

In [10]:
#We will now predict by K nearest Neighbours
predict_1 = []
for i in range(0 , len(data_dist_label)):
     predict_1.append(data_dist_label[i][0])
print(predict_1)

[' Metal', ' Metal', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Metal', ' Metal', ' Metal', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Ceramic', ' Metal', ' Plastic', ' Metal', ' Ceramic', ' Metal', ' Plastic', ' Metal', ' Metal', ' Ceramic', ' Metal', ' Plastic', ' Ceramic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Metal', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Metal', ' Metal', ' Ceramic', ' Metal', ' Plastic', ' Ceramic', ' Ceramic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Metal', ' Ceramic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Metal', ' Plastic', ' Metal', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', '

In [12]:
#We will now find the accuracy for K=1
ps=0 #predicted Successfully
pu=0 #Predicted Unsuccessfully
for i in range(0,len(program_data)):
    if predict_1[i] == program_data[i][3]:
            ps=ps+1
    else :
        pu=pu+1
accuracy_1 = ps/(ps+pu)
print("\nAccuracy for KNN with 1 neighbour")
print("Predicted unsuccesfully  "+str(pu))
print("Predicted Successfully  " + str(ps))
print("Accuracy "+str(accuracy_1*100)+"%")


Accuracy for KNN with 1 neighbour
Predicted unsuccesfully  15
Predicted Successfully  105
Accuracy 87.5%


In [13]:
a = " Metal"
b = " Plastic"
c = " Ceramic"

#We will now intialize the predictions for KNN value 3
predict_3 =[]
for i in range(0 , len(data_dist_label)):
    x=0
    y=0
    z=0
    cols=[]
    for j in range(0 , 3):
        if (str(data_dist_label[i][j]) == a):
            x=x+1
        if(str(data_dist_label[i][j])==b):
            y=y+1
        if(str(data_dist_label[i][j])==c):
            z=z+1
 
    if (x >= y) and (y >= z):
           largest = a
    elif (y >= x) and (y >= z):
        largest = b
    else:
        largest = c
    predict_3.append(largest)
print(predict_3)


[' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Ceramic', ' Metal', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Ceramic', ' Metal', ' Metal', ' Metal', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Metal', ' Plastic', ' Ceramic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Metal', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Ceramic', ' Metal', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Metal', ' Ceramic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Metal', ' Ceramic', ' Plastic

In [15]:
#Now we will find the accuracy for KNN with 3 neighbours

ps=0 #predicted Successfully
pu=0 #Predicted Unsuccessfully
for i in range(0,len(program_data)):
    if predict_3[i] == program_data[i][3]:
            ps=ps+1
    else :
        pu=pu+1
accuracy_1 = ps/(ps+pu)
print("\nAccuracy for KNN with 3 neighbour")
print("Predicted unsuccesfully  "+str(pu))
print("Predicted Successfully  " + str(ps))
print("Accuracy "+str(accuracy_1*100)+"%")


Accuracy for KNN with 3 neighbour
Predicted unsuccesfully  22
Predicted Successfully  98
Accuracy 81.66666666666667%


In [16]:
#We will now intialize the predictions for KNN value 3
predict_5 =[]
for i in range(0 , len(data_dist_label)):
    x=0
    y=0
    z=0
    cols=[]
    for j in range(0 , 5):
        if (str(data_dist_label[i][j]) == a):
            x=x+1
        if(str(data_dist_label[i][j])==b):
            y=y+1
        if(str(data_dist_label[i][j])==c):
            z=z+1
 
    if (x >= y) and (y >= z):
           largest = a
    elif (y >= x) and (y >= z):
        largest = b
    else:
        largest = c
    predict_5.append(largest)
print(predict_5)

[' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Ceramic', ' Metal', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Metal', ' Metal', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Metal', ' Plastic', ' Pla

In [18]:
#Now we will find the accuracy for KNN with 3 neighbours

ps=0 #predicted Successfully
pu=0 #Predicted Unsuccessfully
for i in range(0,len(program_data)):
    if predict_5[i] == program_data[i][3]:
            ps=ps+1
    else :
        pu=pu+1
accuracy_1 = ps/(ps+pu)
print("\nAccuracy for KNN with 5 neighbour")
print("Predicted unsuccesfully  "+str(pu))
print("Predicted Successfully  " + str(ps))
print("Accuracy "+str(accuracy_1*100)+"%")


Accuracy for KNN with 5 neighbour
Predicted unsuccesfully  32
Predicted Successfully  88
Accuracy 73.33333333333333%


As seeing the results are astonishing as we can see that this KNN model has the highest accuracy recorded in this assignment that when N=1 , this KNN model shows accuracy of 87.5%